### PR TITLE
BUG, TST: Check uint64 behaviour in algorithms.py

### PR DESCRIPTION
**Pull request by gfyoung**
First of a series of PR's to patch and test `uint64` behaviour in `core/algorithms.py`.  In this PR, the following functions are checked:

1. `duplicated()` : robust but now has test to confirm
2. `mode()` : robust but now has test to confirm
3. `unique()` : non-robust but patched and tested

### CODE REVIEW

File: pandas/types/common.py [hunk 1]
```diff
@@ -155,6 +155,18 @@ def is_integer_dtype(arr_or_dtype):
             not issubclass(tipo, (np.datetime64, np.timedelta64)))
 
 
+def is_signed_integer_dtype(arr_or_dtype):
+    tipo = _get_dtype_type(arr_or_dtype)
+    return (issubclass(tipo, np.signedinteger) and
```
Comment (jreback) [inline pandas/types/common.py]: ok to add to public api (pandas.api)

File: pandas/core/algorithms.py [hunk 2]
```diff
@@ -508,6 +512,7 @@ def duplicated(values, keep='first'):
     duplicated : ndarray
     """
 
+    values = com._asarray_tuplesafe(values)
```
Comment (jreback) [inline pandas/core/algorithms.py]: hmm, why are you doing this?

File: pandas/core/algorithms.py [hunk 3]
```diff
@@ -548,6 +567,9 @@ def mode(values):
 
     dtype = values.dtype
     if is_integer_dtype(values):
+        # This also works if dtype is uint64 because there is a 1-1
+        # correspondence between int64 and uint64, and we will cast
+        # all elements back to their correct uint64 values.
         values = _ensure_int64(values)
```
Comment (jreback) [inline pandas/core/algorithms.py]: I would rather make this less magical and just select based on dtype. A read will be confused by this; better to do it on dtype.

File: pandas/core/algorithms.py (hunk 2 above)
Comment (gfyoung) [inline pandas/core/algorithms.py]: Your comment <a href="https://github.com/pandas-dev/pandas/pull/14934#discussion_r93349200">below</a> answers this question.  Fair enough.  Will remove.

Comment (gfyoung) [inline pandas/core/algorithms.py]: Yeah, that's perfectly reasonable.  I'll update the documentation to make that clear to avoid confusion.

File: pandas/types/common.py (hunk 1 above)
Comment (gfyoung) [inline pandas/types/common.py]: Yep, done.

**Comment by gfyoung** [comment]
@jreback : Tests are green.  Ready to merge if there is nothing else.

File: pandas/hashtable.pyx [hunk 4]
```diff
@@ -236,6 +236,41 @@ def mode_int64(int64_t[:] values):
 
 @cython.wraparound(False)
 @cython.boundscheck(False)
+def mode_uint64(uint64_t[:] values):
```
Comment (jreback) [inline pandas/hashtable.pyx]: can we move these to tempita? (for mode)

File: pandas/tests/test_algos.py [hunk 5]
```diff
@@ -1202,6 +1209,20 @@ def test_int64_add_overflow():
                                        b_mask=np.array([False, True]))
 
 
+def test_mode_uint64():
```
Comment (jreback) [inline pandas/tests/test_algos.py]: can you add a test class (and consolidate the tests)

File: pandas/hashtable.pyx (hunk 4 above)
Comment (gfyoung) [inline pandas/hashtable.pyx]: Certainly.  Done.

File: pandas/tests/test_algos.py (hunk 5 above)
Comment (gfyoung) [inline pandas/tests/test_algos.py]: Done.

File: pandas/tests/test_algos.py [hunk 6]
```diff
@@ -1202,6 +1209,57 @@ def test_int64_add_overflow():
                                        b_mask=np.array([False, True]))
 
 
+class TestMode(tm.TestCase):
+
+    def test_basic(self):
```
Comment (gfyoung) [inline pandas/tests/test_algos.py]: Done.

**Comment by gfyoung** [comment]
@jreback : Addressed all comments, and everything is green and passing.  Ready to merge if there are no other concerns.

File: pandas/core/algorithms.py [hunk 7]
```diff
@@ -547,10 +566,12 @@ def mode(values):
         constructor = Series
 
     dtype = values.dtype
-    if is_integer_dtype(values):
+    if is_signed_integer_dtype(values):
         values = _ensure_int64(values)
         result = constructor(sorted(htable.mode_int64(values)), dtype=dtype)
```
Comment (jreback) [inline pandas/core/algorithms.py]: side note, not sure whey ``sorted`` is used here, this does a conversion to list (then back when it is constructed).

can you do separate PR to fix this? (use np.sort)

Comment (jreback) [inline pandas/core/algorithms.py]: and it can be at the end of the function (and actually you can move the boiler plate to the end of the function)
(e.g. the construct / sort)

File: pandas/src/hashtable_func_helper.pxi.in [hunk 8]
```diff
@@ -112,3 +112,85 @@ def duplicated_{{dtype}}({{dtype}}_t[:] values,
...
+        int count, max_count = 2
+        int j = -1 # so you can do +=
+        int k
+        kh_{{table_type}}_t *table
+        ndarray[{{ctype}}] modes
+
+    table = kh_init_{{table_type}}()
+
+    {{if dtype == 'object'}}
+    build_count_table_{{dtype}}(values, mask, table)
```
Comment (jreback) [inline pandas/src/hashtable_func_helper.pxi.in]: hmm, any reason NOT to make the signatures the same for build_count_table_*?

File: pandas/core/algorithms.py (hunk 7 above)
Comment (jreback) [inline pandas/core/algorithms.py]: or can you do in this PR

**Comment by jreback** [comment]
few comments

File: pandas/src/hashtable_func_helper.pxi.in (hunk 8 above)
Comment (gfyoung) [inline pandas/src/hashtable_func_helper.pxi.in]: `build_count_table_object` takes a `mask` argument, whereas the others do `null` detection internally and take a `dropna` parameter instead.  It might be possible to standardize, but as all of these functions are tied to Python-facing functions (e.g. `value_counts`), I don't think it would be a good idea to investigate in this PR.

File: pandas/core/algorithms.py (hunk 7 above)
Comment (gfyoung) [inline pandas/core/algorithms.py]: And how is that any different from `np.sort`, which converts to `ndarray` before we then convert to `Series`?  We might as well just use `Series.sort_values` then?

Comment (jreback) [inline pandas/core/algorithms.py]: its not, yes you could do a ``.sort_values()`` (I just don't want ``sorted``)

File: pandas/src/hashtable_func_helper.pxi.in (hunk 8 above)
Comment (jreback) [inline pandas/src/hashtable_func_helper.pxi.in]: hmm, ok, can you make an issue about this then? 

this PR still need some work, so might be simpler to do that first FYI.

File: pandas/core/algorithms.py (hunk 7 above)
Comment (gfyoung) [inline pandas/core/algorithms.py]: Can't quite move to the end (`categorical` doesn't sort).  However, can just append to the end of the `Series` construction instead.

File: pandas/src/hashtable_func_helper.pxi.in (hunk 8 above)
Comment (gfyoung) [inline pandas/src/hashtable_func_helper.pxi.in]: Actually, it doesn't seem as difficult as I thought previous.  I'm going to try to make the change here.

Comment (jreback) [inline pandas/src/hashtable_func_helper.pxi.in]: ok, lmk then.

Comment (gfyoung) [inline pandas/src/hashtable_func_helper.pxi.in]: So looking into this further, I think I understand why `object` was special-cased from the rest.  It appears that Python's hashing for `object` (`PyObject_Hash`) does not work very well with `null` values.  Thus, when you insert a value like `nan` into such a hashtable multiple times, they will be interpreted as different keys, which is obviously wrong.

In light of this, I am inclined to leave this code alone.  Thoughts?

**Comment by jreback** [comment]
thanks!
